### DISCUSSION TITLE
Generalized Hill criterion

### DISCUSSION THREAD

**Discussion by Max-EDFLMT** [thelfer/tfel#130] [Q&A]
Hi,

I'm actually using a version of the Hill criterion available in the MFront gallery that I would like to improve.
In this model, the equivalent stress is computed using the Hill tensor computed by the correctly named function @HillTensor.
Would it be possible to expand this to the generalized Hill criterion where the exponents of the yield function is not 2 but an integer n ?

I volunteer to test the different algorithms and see if everything is working fine at high values of n if the implementation in itself is "easy" !

Thanks in advance,
Maxime

**Comment by thelfer** [comment]
@Max-EDFLMT We already have the Barlat stress criterion, see <https://thelfer.github.io/tfel/web/barlat-yld2004.html>.

Following the work of Barlat, we could also define a linear transformation of the stress to transform the isotropic Hosford stress criterion (see <https://thelfer.github.io/tfel/web/hosford.html>) to an orthotropic stress criterion. 

Would one of two options match your need ?

**Reply by Max-EDFLMT** [reply]
Thanks for the quick answer.
The generalized Hill criterion is written as in :
https://en.wikipedia.org/wiki/Hill_yield_criterion#Generalized_Hill_yield_criterion

If I understood you clearly, this is similar to defining the linear transformation of the stress in the Hosford criterion. Though, I'm not sure about the effect of the exponent n after that.

**Reply by thelfer** [reply]
In this isotropic case, the n exponent of the Hosford criterion allows to smoothly transform the von Mises yield surface to the Tresca yield surface.

I never worked with the generalized Hill criterion right now. Is there an expression of this criterion not based on the principal stress (like the standard Hill criterion).

**Reply by thelfer** [reply]
I know that Barlat and coworkers were able to derive closed form expression of the Barlat and Hosford stress criterion based on invariants for integer values of `n`. This is very convenient for an implementation point of view..

**Reply by Max-EDFLMT** [reply]
I don't know if such expressions exist. A quick look in the literature gave me nothing, I guess I will need to go deeper !

**Reply by Max-EDFLMT** [reply]
The original article from Hill (1979) "Theoretical plasticity of textured aggregates", gives not such expressions like it was the case for the standard Hill (1949) criterion.

**Reply by thelfer** [reply]
It would be nice to avoid working with principal stresses as the derivatives can become singular. We handle that for the Hosford case, but it's pretty painfull.

**Reply by Max-EDFLMT** [reply]
In order not to perform painful operations on your side, would it be possible to apply directly the linear transformation to the Hosford criterion if the developments are already made for it ?

**Reply by thelfer** [reply]
I think so. One needs to write the derivatives and see what happens.

**Reply by Max-EDFLMT** [reply]
I will try Barlat to see if it fits my needs. From what little I know already, it should be equivalent in some extent, choosing the right parameters. They are not easily identified given the criterion formulation, though

**Reply by thelfer** [reply]
Barlat uses two linear transformations. Using only one reduces the number of parameters significantly.

**Reply by thelfer** [reply]
Hi @Max-EDFLMT,

Please find attached a first implementation. I have done this for fun, so no guarantees. You probably want to read this tutorial closely to improve the robustness of the implementation: <https://thelfer.github.io/tfel/web/hosford.html>.

I used coefficients used by Barlat and it leads to the following yield surface in the pi-plane:

![pi-plane](https://user-images.githubusercontent.com/13975528/156895549-b1a0a868-fc26-4939-8318-b86a57a1641c.png)

Let me know if it fits your needs.

[OrthotropicHosford.zip](https://github.com/thelfer/tfel/files/8191221/OrthotropicHosford.zip)

**Reply by Max-EDFLMT** [reply]
Hi Thomas,
This seems to be really suitable !
I will try to fit it on my data but this should work.

And sorry for my late response, I was on holidays.

Maxime

PS: that projectOnPiPlane function is really convenient ;-)

**Reply by thelfer** [reply]
Let me know if you make any progress.

**Reply by Max-EDFLMT** [reply]
I'm facing an issue when building the library

```
Treating target : all
In file included from asterOrthotropicHosford.cxx:11:0:
OrthotropicHosford.mfront: In member function ‘bool tfel::material::OrthotropicHosford<hypothesis, Type, false>::computeFdF(bool)’:
OrthotropicHosford.mfront:60:12: error: expected unqualified-id before ‘[’ token
compilation terminated due to -Wfatal-errors.
Makefile.mfront:37 : la recette pour la cible « asterOrthotropicHosford.o » a échouée
make: *** [asterOrthotropicHosford.o] Erreur 1
callMake: can't build target 'all'
libraries building went wrong
Command was: make -C src -f Makefile.mfront all -s 
```

Is this a version issue ? I'm trying to install 4.0 but it is more complicated than I believed.

I'm actually using mfront-3.4.0 on a scibian station.

**Reply by Max-EDFLMT** [reply]
Figured out myself that it was a C++ issue, I'm moving forward and i'll let you know if it works

**Reply by thelfer** [reply]
@Max-EDFLMT I used TFEL-4.0 (C++-17), but it can be backported to TFEL-3.4 (C++-11). Let me know if you have difficulties.

**Reply by Max-EDFLMT** [reply]
I found the conversion to the former C++ version. However I'm struggling with installing mfront on my scibian station.
I found a way around by using the salome_meca environment and specifying the libraries paths inside the main python pi-plane.py file. This seem highly cumbersome but, at least, it works. I managed to reproduce your results ! I'll let you know for further progress.
Thanks for your reactivity !

**Reply by thelfer** [reply]
@Max-EDFLMT Do not hesitate to open an issue regarding the installation on scibian (which is a modified debian version, isn't it ?).
